src/consensus/produce_block.rs

```diff
@@ -126,45 +112,46 @@ where
             parent_slot,
         );
 
-        let mut sleep_duration = DELTA_BLOCK;
-        for i in 0.. {
-            let slice_index = if i == 0 {
-                Either::Left((parent_block_id, 0))
+        let mut duration_left = DELTA_BLOCK;
+        for slice_index in 0.. {
+            let parent = if slice_index == 0 {
+                Some(parent_block_id)
             } else {
-                Either::Right(NonZeroUsize::new(i).unwrap())
+                None
             };
-            let (slice, cont_prod) =
-                produce_slice(&self.txs_receiver, slot, slice_index, sleep_duration).await;
-            if i == 0 {
-                assert!(slice.parent.is_some());
-            }
+            let (payload, maybe_duration) =
+                produce_slice_payload(&self.txs_receiver, parent, duration_left).await;
 
-            let is_last = slice.is_last;
-            let maybe_block_hash = self.shred_and_disseminate(slice).await?;
-            match cont_prod {
-                Continue::Stop => {
-                    assert!(is_last);
-                    let block_hash = maybe_block_hash.unwrap();
-                    return Ok((slot, block_hash));
-                }
-                Continue::Continue { left } => {
-                    assert!(!is_last);
-                    assert!(maybe_block_hash.is_none());
-                    sleep_duration = left;
-                }
+            match self
+                .shred_and_disseminate(slice_index, slot, payload, maybe_duration)
+                .await?
+            {
+                Either::Left(block_hash) => return Ok((slot, block_hash)),
+                Either::Right(duration) => duration_left = duration,
             }
         }
         unreachable!()
     }
 
-    /// Shreds and disseminates the slice.
+    /// Shreds and disseminates the slice payload.
     ///
-    /// Returns Ok(Some(Hash)) if slice.is_last == true.
-    /// Returns Ok(None) if slice.is_last == false.
-    async fn shred_and_disseminate(&self, slice: Slice) -> Result<Option<Hash>> {
+    /// Returns Ok(Either::Left(hash of the block)) if this is the last slice.
+    /// Returns Ok(Either::Right(duration left in block)) if this is not the last slice.
+    async fn shred_and_disseminate(
+        &self,
+        slice_index: usize,
+        slot: Slot,
```

Comment:
nit: I would switch `slot` and `slice_index`, as that's the order we always use (from larger unit to smaller, e.g. epoch, slot, slice, shred)

Reply:
I suppose based on the comment below, this will become obsolete.